### PR TITLE
Associations

### PR DESCRIPTION
**Pull request by iammouaz**
Branch changes : 

- Added Classroom class
- Implemented association classroom to-many students
- Implemented association a student to only one classroom
- Added Book and Rental classes
- Implemented association book to-many rentals
- Implemented association rental to only one book
- Implemented association person to-many rentals
- Implemented association rental to only one person
- Modified other classes to work with the new structure

### CODE REVIEW

**Review by ZeenatLawal — COMMENTED** [review]
Hi @iammouaz, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request ❌ 

You have submitted a project, that has no README file.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **Does not count** into the code reviews limit._

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  attr_reader :students
+
+  def initialize(label)
+    @label = label
+  end
+
+  def add_student(student)
+    @students.push(student)
+    student.classroom = self
+  end
```
Comment (ZeenatLawal) [inline classroom.rb:13]: - [x] Kindly create a belongs-to relationship between `Classroom` and `Student` (a student belongs to a classroom) in the Student Class. You can refer to this [article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships) for examples.

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+class Book
+  attr_accessor :title, :author, :rentals
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+end
```
Comment (ZeenatLawal) [inline book.rb]: - [x] The `Book` class should have a method for adding rentals. The method should take in two parameters (`person` and `date`) and return a Rental instance that takes in `date`, `person` and, `self` as parameters.

File: person.rb [hunk 3]
```diff
@@ -1,10 +1,18 @@
 require_relative 'corrector'
+require './rental'
 
 class Person
   attr_accessor :name, :age
   attr_reader :id
```
Comment (ZeenatLawal) [inline person.rb]: - [x] There should be getter and setter (attr_accessor) for :rentals. Kindly add this.

File: person.rb [hunk 4]
```diff
@@ -1,10 +1,18 @@
...
 
-  def initialize(age: 11, name: 'Unknown', parent_permission: true); end
+  def initialize(age: 11, name: 'Unknown', parent_permission: true)
+    @id = id
+    @name = name
+    @age = age
+    @parent_permission = parent_permission
+    @corrector = Corrector.new
+    @rentals = []
+  end
```
Comment (ZeenatLawal) [inline person.rb:15]: - [x] The `Person` class should have a method for adding rentals. The method should take in two parameters (`book` and `date`) and return a Rental instance that takes in `date`, `book` and, `self` as parameters.

**Review by ZeenatLawal — CHANGES_REQUESTED** [review]
Hi @iammouaz, 

Great job so far!:clap: :clap:
## To Highlight 🎊 🎊 
- Git flow is followed. ✔️
- PR has a title and summary. ✔️

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: classroom.rb (hunk 1 above)
Comment (ZeenatLawal) [reply classroom.rb:13]: - [ ] The issue here has not been properly resolved. The changes are to be made in the Student class, kindly create the belongs-to relationship in the Student class. Example shown below, in this case `owner` should be `classroom` and `animals` should be `students`.

```ruby
  def owner=(owner)
    @owner = owner
    owner.animals.push(self) unless owner.animals.include?(self)
  end
```

File: book.rb (hunk 2 above)
Comment (ZeenatLawal) [reply book.rb]: - [ ] The issue here has been properly resolved

File: person.rb (hunk 3 above)
Comment (ZeenatLawal) [reply person.rb]: - [ ] The issue here has been properly resolved

File: person.rb (hunk 4 above)
Comment (ZeenatLawal) [reply person.rb:15]: - [ ] The issue here has been properly resolved

File: classroom.rb [hunk 5]
```diff
@@ -4,6 +4,7 @@ class Classroom
 
   def initialize(label)
     @label = label
+    @students = []
```
Comment (ZeenatLawal) [inline classroom.rb:7]: - [ ] Kindly remove this line of code and review the comment made above.

**Review by ZeenatLawal — CHANGES_REQUESTED** [review]
Hi @iammouaz, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
- [ ] _Kindly check the comments above, made under the reviews of the previous code reviewer(s) and stating **not properly resolved**. Kindly take your time to correctly review those issues outlined by the previous code reviewer, and please check to see that the changes are working as intended._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: classroom.rb (hunk 5 above)
Comment (iammouaz) [reply classroom.rb:7]: @ZeenatLawal I don't understand... so I only want to do is removing the @students line ??

**Review by eliasecasta — COMMENTED** [review]
Hello @iammouaz 👋,

_This is an additional review as a result of the second opinion request._

I do agree with @ZeenatLawal about the change that she required you to apply in your students class which will be necessary for the next milestones, this is just to help you get prepared and all set for them later on and avoid you bugs, please implement the change suggested in your Student class to comply with the requirements by microverse provided to us for this milestone.

I do however disagree with Zeenat about the removal of the students' array in the classroom class and the removal of the push of @students.push(student) in your add_student method, it will be necessary for the next milestones, please add again that lines of code because they were correct and will avoid you issues in the next milestones.

@ZeenatLawal, please don't forget to use inline comments specifically where you are requesting the changes to avoid confusion.

Please make these two changes so you can get approved your project in the next review, good luck and happy coding!

Keep rocking 🚀!

**Comment by ZeenatLawal** [comment]
Hello @iammouaz and @e71az, apologies about the inline comments confusion and the code removal comment will take note of that next time. @e71az thanks for the corrections.

**Review by AshakaE — APPROVED** [review]
### Hi @iammouaz   🙋‍♀️, 

# STATUS APPROVED ✔ 
   
Your project is complete!  There is nothing else to say other than... it's time to merge it 🔗 🚢 
Congratulations! 🎉 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers 🥂  and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Dipeshtwis** [comment]
# Status: Invalid ❌

Hi @iammouaz,

While you made a great effort in this project, unfortunately, I cannot proceed to review your code. 

Please fix the issue mentioned below and then request for code review

## Reason

- [ ] You have requested code review which was already approved. If you are repeating the week, Please make sure to use a new PR or a new repository with all the changes that need to be there in this milestone.

### Please make sure to submit code review when you are done with the changes that was requested

**Don't worry you still have same number of requests left as previous.**


Cheers and Happy Coding :clap: :clap: :clap: 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Dipesh Kumar_**

**Review by peterrobert — COMMENTED** [review]
### INVALID 🛑  
Hello @iammouaz ,

- [ ] It seems you have submitted an already reviewed and merged pull request. For this reason, I can not review your project. Kindly submit the correct pull request. Thanks

**Comment by rukundoeric** [comment]
Hi @iammouaz, 

_While you made a great effort in this project, unfortunately, I cannot proceed to review your code._

## Invalid Code Review Request ⚠️

- You have submitted a PR that is already reviewed, approved, and merged. if there is any specific reason for that, kindly request another review explaining why.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

>Invalid Code Review Request **does not count** into the code reviews limit.

**Review by Ghiftee — COMMENTED** [review]
## STATUS: INVALID ❌ ❌ 
Hi @iammouaz, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that has already been approved and merged. According to the additional information you provided, it says you repeated the week. I am really sorry about that however you still need to create a new PR for this project.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Pull request merged**